File: 22. Generate Parentheses.cpp
class Solution {
private:
    vector <string> res;
public:
    vector<string> generateParenthesis(int n) {
        DFS(n, n, n, "");
        return res;
    }
    
    void DFS(int left, int right, int n, string path)
    {
        if (path.length() == 2 * n)
        {
            res.push_back(path);
            return;
        }
        if (left == right)
            DFS(left-1, right, n, path + "(");
        else if (left < right)
        {
            for (int i = 0; i < 2; i++)
            {
                if (i == 0)
                    DFS(left, right-1, n, path + ")");
                else
                {
                    if (left > 0)
                        DFS(left-1, right, n, path + "(");
                }
            }
        }
    }
};

File: 22. Generate Parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        self.res = []
        left, right = n, n
        self.DFS(left, right, n, "")
        return self.res
        
    def DFS(self, left, right, n, path):
        if len(path) == 2*n:
            self.res += [path]
            return
        if left == right:
            self.DFS(left-1, right, n, path + '(')
        elif left < right:
            for i in xrange(2):
                if i == 0:
                    self.DFS(left, right-1, n, path + ')')
                else:
                    if left > 0:
                        self.DFS(left-1, right, n, path + '(')
